File: backend/app/schemas/security_analytics.py
"""
Schemas for the Security Analytics endpoints.
"""
from datetime import datetime
from typing import Any, Dict, List, Optional
from pydantic import BaseModel


# ─── ThreatEvent ──────────────────────────────────────────────────────────────

class ThreatEventOut(BaseModel):
    id: int
    ip_address: str
    event_type: str
    source: str
    severity: str
    score: float
    metadata_json: Optional[str] = None
    created_at: datetime

    class Config:
        from_attributes = True


# ─── ThreatScore ──────────────────────────────────────────────────────────────

class ThreatScoreOut(BaseModel):
    ip_address: str
    current_score: float
    threat_level: str
    last_updated: datetime
    active_block_id: Optional[int] = None

    class Config:
        from_attributes = True


# ─── Threat detail (score + event timeline) ────────────────────────────────────

class ThreatDetail(BaseModel):
    score: ThreatScoreOut
    events: List[ThreatEventOut]


# ─── Dashboard metrics ─────────────────────────────────────────────────────────

class SecurityMetrics(BaseModel):
    active_blocks: int
    critical_threats: int
    honeypot_hits_24h: int
    failed_logins_24h: int
    average_threat_score: float
    top_offending_ips: List[Dict[str, Any]]


# ─── Recalculate response ──────────────────────────────────────────────────────

class RecalculateResponse(BaseModel):
    recalculated: int
    message: str


# ─── Block Management ──────────────────────────────────────────────────────────

class ThreatBlockOut(BaseModel):
    id: int
    ip_address: str
    reason: Optional[str] = None
    threat_score: int
    hit_count: int
    first_seen: datetime
    last_seen: datetime
    blocked_at: Optional[datetime] = None
    expires_at: Optional[datetime] = None
    status: str

    class Config:
        from_attributes = True

class ManualBlockRequest(BaseModel):
    reason: Optional[str] = "Manual admin block"
    duration_minutes: Optional[int] = 60

File: frontend/src/api/securityAnalytics.ts
import api from "./client";

export interface ThreatEvent {
  id: number;
  ip_address: string;
  event_type: string;
  source: string;
  severity: string;
  score: number;
  metadata_json?: string;
  created_at: string;
}

export interface ThreatScore {
  ip_address: string;
  current_score: number;
  threat_level: string;
  last_updated: string;
  active_block_id?: number;
}

export interface ThreatDetail {
  score: ThreatScore;
  events: ThreatEvent[];
}

export interface SecurityMetrics {
  active_blocks: number;
  critical_threats: number;
  honeypot_hits_24h: number;
  failed_logins_24h: number;
  average_threat_score: number;
  top_offending_ips: {
    ip_address: string;
    current_score: number;
    threat_level: string;
  }[];
}

export interface RecalculateResponse {
  recalculated: number;
  message: string;
}

export interface ThreatBlock {
  id: number;
  ip_address: string;
  reason?: string;
  threat_score: number;
  hit_count: number;
  first_seen: string;
  last_seen: string;
  blocked_at?: string;
  expires_at?: string;
  status: string;
}

export const getThreats = async (): Promise<ThreatScore[]> => {
  const response = await api.get("/api/v1/security/threats");
  return response.data;
};

export const getThreatDetail = async (ipAddress: string): Promise<ThreatDetail> => {
  const response = await api.get(`/api/v1/security/threats/${ipAddress}`);
  return response.data;
};

export const getSecurityMetrics = async (): Promise<SecurityMetrics> => {
  const response = await api.get("/api/v1/security/metrics");
  return response.data;
};

export const recalculateScores = async (): Promise<RecalculateResponse> => {
  const response = await api.post("/api/v1/security/recalculate");
  return response.data;
};

export const getBlocks = async (): Promise<ThreatBlock[]> => {
  const response = await api.get("/api/v1/security/blocks");
  return response.data;
};

export const manualBlockIp = async (ipAddress: string, reason?: string, durationMinutes = 60): Promise<ThreatBlock> => {
  const response = await api.post(`/api/v1/security/blocks/${ipAddress}/block`, {
    reason,
    duration_minutes: durationMinutes,
  });
  return response.data;
};

export const manualUnblockIp = async (ipAddress: string): Promise<ThreatBlock> => {
  const response = await api.post(`/api/v1/security/blocks/${ipAddress}/unblock`);
  return response.data;
};
